File: python/src/ParserState.py
from src.Settings import LangData
from src.DomNode import DomNode
from src.FunctionNode import FunctionNode

class ParserState:
  def __init__(self, parser):
    self.parser = parser

class ParserStateDefault(ParserState):
  def parseToken(self, token):
    if token.type == 'TagOpen':
      self.parser.state = ParserStateTag(self.parser)
    elif token.type == 'TagNameScoped' or token.type == 'TagNameScopeless':
      node = DomNode(token.text[1:], token.type == 'TagNameScopeless')
      self.parser.currentScope.appendChild(node)
      self.parser.lastTag = node
    elif token.type == 'FunctionName':
      func = FunctionNode(token.text[1:])
      func.setParent(self.parser.currentScope)
      self.parser.funcCalls.append(func)
      self.parser.lastTag = func
    elif token.type == 'ScopeOpen' or token.type == 'ScopeClose':
      self.parser.state = ParserStateScope(self.parser)
      self.parser.state.parseToken(token)
    else:
      self.parser.currentScope.appendText(token.text)

class ParserStateTag(ParserState):
  def __init__(self, parser):
    super().__init__(parser)
    self.node = self.parser.lastTag
    self.index = 0
    self.lastParamKey = ''
  def parseToken(self, token):
    if token.type == 'TagClose':
      self.parser.state = ParserStateDefault(self.parser)
    elif token.type == 'TagId':
      self.node.parameters['id'] = token.text[1:]
    elif token.type == 'TagClass':
      try: self.node.parameters['class'] += ' '+token.text[1:]
      except KeyError: self.node.parameters['class'] = token.text[1:]
    elif token.type == 'TagFlagParam':
      self.node.parameters[token.text[1:]] = 'true'
    else:
      self.index += 1
      if self.index % 2: self.lastParamKey = token.text
      else: self.node.addParameter(self.lastParamKey, token.text)

class ParserStateScope(ParserState):
  def parseToken(self, token):
    if token.type == 'ScopeOpen':
      self.parser.currentScope = self.parser.lastTag
    elif token.type == 'ScopeClose':
      self.parser.currentScope = self.parser.currentScope.parent
    self.parser.state = ParserStateDefault(self.parser)

File: javascript/src/ParserState.js

class ParserState {
  constructor(parser) {
    this.parser = parser;
  }
}

class ParserStateDefault extends ParserState {
  parseToken(token) {
    if (token.type == 'TagOpen') {
      this.parser.state = new ParserStateTag(this.parser);
    } else if (['TagNameScoped', 'TagNameScopeless'].includes(token.type)) {
      let node = new ShpNode(token.text.substr(1));
      this.parser.currentScope.appendChild(node);
      this.parser.lastNode = node;
    } else if (['ScopeOpen', 'ScopeClose'].includes(token.type)) {
      this.parser.state = new ParserStateScope(this.parser);
      this.parser.state.parseToken(token);
    } else {
      this.parser.currentScope.appendChild(new ShpNodeText(token.text + ' '));
    }
  }
}

class ParserStateTag extends ParserState {
  constructor(parser) {
    super(parser);
    this.node = this.parser.lastNode;
    this.index = 0;
    this.lastParamKey = '';
  }
  parseToken(token) {
    if (token.type == 'TagClose') {
      this.parser.state = new ParserStateDefault(this.parser);
    } else if (token.type == 'TagId') {
      this.node.addParameter('id', token.text.substr(1));
    } else if (token.type == 'TagClass') {
      this.node.addParameter('class', token.text.substr(1));
    } else if (token.type == 'TagFlagParam') {
      this.node.addParameter(token.text.substr(1), 'true');
    } else {
      this.index += 1;
      if (this.index % 2) this.lastParamKey = token.text;
      else this.node.addParameter(this.lastParamKey, token.text);
    }
  }
}

class ParserStateScope extends ParserState {
  parseToken(token) {
    if (token.type == 'ScopeOpen') {
      this.parser.currentScope = this.parser.lastNode;
    } else if (token.type == 'ScopeClose') {
      this.parser.currentScope = this.parser.currentScope.parent;
    }
    this.parser.state = new ParserStateDefault(this.parser);
  }
}
